tests/cli/test_vscode_utils.py

```diff
@@ -39,3 +52,270 @@ def test_get_expected_vscode_settings_json_path():
+        _make_fake_vscode_extension(
+            vscode_extensions_dir=extensions_dir,
+            extension_id=extension_id,
+            version=version,
+        )
+
+
+def _fake_vscode_extension_json_entry(
+    extension_id: str = "ms-python-python", version: str = "2023.22.1"
+) -> dict[str, Any]:
+    return {
+        "identifier": {
+            "id": extension_id,
+            "uuid": "f1f59ae4-9318-4f3c-a9b5-81b2eaa5f8a5",
+        },
+        "version": "2023.22.1",
+        "location": {
+            "$mid": 1,
+            # TODO: Check if we need to update this path in each entry once we copy
+            # the extension folders to the remote, or if this gets automatically set
+            # to the right path using the `relativeLocation` field.
+            "path": f"/home/fabrice/.vscode/extensions/{extension_id}-{version}",
+            "scheme": "file",
+        },
+        "relativeLocation": f"{extension_id}-{version}",
+        "metadata": {
+            "id": "f1f59ae4-9318-4f3c-a9b5-81b2eaa5f8a5",
+            "publisherId": "998b010b-e2af-44a5-a6cd-0b5fd3b9b6f8",
+            "publisherDisplayName": "Microsoft",
+            "targetPlatform": "undefined",
+            "isApplicationScoped": False,
+            "updated": True,
+            "isPreReleaseVersion": False,
+            "installedTimestamp": 1702999774872,
+            "preRelease": False,
+        },
+    }
+
+
+def _make_fake_vscode_extension(
+    vscode_extensions_dir: Path,
+    extension_id: str = "ms-python-python",
+    version: str = "2023.22.1",
+) -> Path:
+    extension_dir = vscode_extensions_dir / f"{extension_id}-{version}"
+    extension_dir.mkdir()
+
+    (extension_dir / "package.json").write_text(
+        json.dumps(
+            {
+                "name": "python",
+                "displayName": "Python",
+                "description": "...",
+                "version": f"{version}",
+                "foo": "bar",
+            }
+        )
+    )
+    # Add some dummy files to the extension directory.
+    (extension_dir / "pythonExtensionApi" / "src").mkdir(parents=True)
+    (extension_dir / "pythonExtensionApi" / "src" / "bob.txt").write_text("foo")
+    return extension_dir
+
+
+def assert_files_were_copied_correctly(source: Path, dest: Path):
+    """If `source` is a dir, checks that all files in `source` have been copied to
+    `dest`.
+
+    If `source` is a file, checks that the content of `dest` is the same.
+    """
+    assert source.exists()
+    if source.is_dir():
+        assert dest.is_dir()
+        for file in source.iterdir():
+            assert_files_were_copied_correctly(file, dest / file.name)
+    else:
+        assert dest.exists()
+        assert not dest.is_dir()
+
+        assert source.stat().st_mode == dest.stat().st_mode
+        assert source.stat().st_size == dest.stat().st_size
+        # assert source.read_text() == dest.read_text()
+        assert source.read_bytes() == dest.read_bytes()
```

Comment:
Could probably be simplified with a `diff --recursive` then parse the stdout of the command no?

Reply:
I'd like to keep things in python as much as possible, also so that tests work on Windows.

Reply:
Fixed in e1512f1e776fae292a6ac976547f8648c1a2b9ae using the `filecmp` module.